### PR TITLE
Hide simulator_state

### PR DESCRIPTION
**Pull request by dabacon**
The simulator state for simulators is an internal state used by the implementations for simulating state when stepping through a simulation.  It shouldn't be exposed "publically".  See #1658 for where callers where using this state instead of going to the public methods on step result or trial result.

### CODE REVIEW

**Comment by dabacon** [comment]
Fixes #1658

File: cirq/sim/simulator.py [hunk 1]
```diff
@@ -397,8 +397,12 @@ def __init__(self,
         self.measurements = measurements or collections.defaultdict(list)
 
     @abc.abstractmethod
-    def simulator_state(self) -> Any:
-        """Returns the simulator_state of the simulator after this step.
+    def _simulator_state(self) -> Any:
+        """Returns the simulator state of the simulator after this step.
+
+        This method starts with an underscore to indicate that it is private.
+        State that should be exposed should exposed on the child StepResult
```
Comment (viathor) [inline cirq/sim/simulator.py]: nit: This is a little convoluted. Maybe "Instead, use public state exposed by StepResult.".

File: cirq/sim/density_matrix_simulator.py [hunk 2]
```diff
@@ -618,10 +619,10 @@ def __init__(self,
...
                         sorted(self.measurements.items())}
-        return (self.params, measurements, self.final_simulator_state)
+        return (self.params, measurements, self._final_simulator_state)
 
     def __repr__(self):
         return ("cirq.DensityMatrixTrialResult(params={!r}, measurements={!r}, "
-                "final_simulator_state={!r})"
-                .format(self.params, self.measurements,
-                        self.final_simulator_state))
+                "final_simulator_state={!r})".format(
```
Comment (Strilanc) [inline cirq/sim/density_matrix_simulator.py]: If we don't want to expose the final simulator state, why are we including it in the trial results objects and printing it out?

Comment (dabacon) [reply cirq/sim/density_matrix_simulator.py]: This is a repr. We want it to generate the code needed to create the object, no?  If it were __str__ I'd agree.

File: cirq/sim/simulator.py (hunk 1 above)
Comment (dabacon) [reply cirq/sim/simulator.py]: done

**Comment by mpharrigan** [comment]
I was using `qubit_map` to pass to `cirq.pauli_string_expectation`. Seems sortof important to have this tagging along (in a public way) when implicitly labeling qubits by e.g. row index. Is there a preferred workaround?

**Comment by dabacon** [comment]
we should expose it in the result methods.

**Pull request merged**